single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+
+As a result, it seems we can use the `distribution` project from docker.
+
+## Implementation notes
+
+Very short and brief implementation idea. We can plan more details once we agree on the idea and solution.
+To keep the migration simpler, we split it separately for the container images and for the helm charts cache.
+We want to start with the images cache, as that's what we currently use Docker Hub for.
+
+### Container images
+
+Currently, we use two ACR registries, `giantswarm` for container images and `giantswarmpublic` for helm charts.
+The images one is already over 4 TB in size and it will be hard to clean it up. We can instead start with a new
+empty repo and populate it using a separate instance of `retagger`. We won't start uploading build artifacts there
+before we figure out how to use the retention policy.
+
+High level migration plan:
+
+- bootstrap a new ACR registry for images, let's call it tentatively `gsoci`
+  - configure access, geo-replication and a retention policy for the repo
+- prepare and deploy `distribution` cache instance, configure it for the new repo
+  - ensure monitoring and alerting
+- switch our CI/CD pipelines to not tag dev build artifacts
+  - use a tool (`skopeo`, `retagger`?) to replicate all (do we really need all?) the tagged versions of images we build from the old `giantswarm` repo to the new `gsoci`
+  - configure CI/CD so it uploads to both the old `giantswarm` and the new `gsoci`
+- deploy a new instance of `retagger` that will replicate all the configured public images to the new `gsoci` registry
+- switch container runtimes to use the cache as a source of images and upstream `gsoci` as a fallback
+- cleanup (after some time)
+  - delete our registries in docker hub and quay
+  - delete the old `giantswarm` registry in ACR
+- for this part, we keep China registry unchanged
+  - we reconfigure the `crsync` to replicate images from `gsoci` to `aliyun` in China
+
+#### China extension
+
+- we try to setup automated geo-replication between Europe and China
+- switch China clusters to use the new ACR-China registry
+- shut down `crsync` entirely
+- delete Aliyun registry
+
+### Helm charts
+
+The blocker for switching Helm Charts entirely to OCI is currently our artifact discovery process, which uses the
+`index.yaml` file, that is available in HTTP based Helm registries only. So, to get started, we have to switch our
+discovery process to OCI based one. Currently, we're already uploading our build artifacts to both HTTP and OCI repos,
+so nothing is blocking implementation and a switch.
+
+High level migration plan:
+
+- prepare and deploy `distribution` cache instance, configure it for the upstream ACR registry
+  - ensure monitoring and alerting
+- switch our CI/CD pipelines to not tag dev build artifacts
+  - we want to use retention policies for helm charts as well
+- implement changes in the `app-operator`
+  - use OCI discovery instead of `index.yaml` from HTTP registries
```

Comment:
How does this -if so - affect creating Flux HelmRepositories instead of Catalogs?

Reply:
I think it doesn't - [`HelmRepository` supports OCI just as a type config option](https://fluxcd.io/flux/components/source/helmrepositories/#type) and we should already have this available, as we already support installing stuff from OCI registries. So, I expect changes done by Łukasz cover this already, but can you confirm @ljakimczuk ?